### PR TITLE
mssql_db: add support for transparent data encryption

### PR DESCRIPTION
**Pull request by koikonom**
TDE is enabled by default for some SKUs but not for others, so we cannot
just introduce this setting.

Since this is a breaking change the setting will be set behind a feature
flag.

### CODE REVIEW

File: internal/services/mssql/mssql_database_resource.go [hunk 1]
```diff
@@ -759,6 +799,20 @@ func resourceMsSqlDatabaseRead(d *pluginsdk.ResourceData, meta interface{}) erro
 		return fmt.Errorf("setting `geo_backup_enabled`: %+v", err)
 	}
 
+	if features.ThreePointOh() {
+		tde, err := transparentEncryptionClient.Get(ctx, id.ResourceGroup, id.ServerName, id.Name)
+		if err != nil {
+			return fmt.Errorf("while retrieving Transparent Data Encryption status of %q: %+v", id.String(), err)
+		}
+		tdeStatus := false
+		if tde.TransparentDataEncryptionProperties.Status == sql.TransparentDataEncryptionStatusEnabled {
```
Comment (tombuildsstuff) [inline internal/services/mssql/mssql_database_resource.go:808]: there's a potential crash here if it's disabled:

```suggestion
		if tde.TransparentDataEncryptionProperties != nil && tde.TransparentDataEncryptionProperties.Status == sql.TransparentDataEncryptionStatusEnabled {
```

File: internal/services/mssql/mssql_database_resource.go [hunk 2]
```diff
@@ -759,6 +799,20 @@ func resourceMsSqlDatabaseRead(d *pluginsdk.ResourceData, meta interface{}) erro
...
+			tdeStatus = true
+		}
+		if err := d.Set("transparent_data_encryption_enabled", tdeStatus); err != nil {
+			return fmt.Errorf("setting transparent_data_encryption: %+v", err)
+		}
```
Comment (tombuildsstuff) [inline internal/services/mssql/mssql_database_resource.go:813]: this is a bool so there's no need to error check those?

```suggestion
		d.Set("transparent_data_encryption_enabled", tdeStatus)
```

File: internal/services/mssql/mssql_database_resource.go [hunk 3]
```diff
@@ -558,6 +581,22 @@ func resourceMsSqlDatabaseCreateUpdate(d *pluginsdk.ResourceData, meta interface
 		return fmt.Errorf("waiting for create/update of %s: %+v", id, err)
 	}
 
+	if features.ThreePointOh() {
+		statusProperty := sql.TransparentDataEncryptionStatusDisabled
+		encryptionStatus := d.Get("transparent_data_encryption_enabled").(bool)
+		if encryptionStatus {
+			statusProperty = sql.TransparentDataEncryptionStatusEnabled
+		}
+		_, err := transparentEncryptionClient.CreateOrUpdate(ctx, id.ResourceGroup, id.ServerName, id.Name, sql.TransparentDataEncryption{
```
Comment (tombuildsstuff) [inline internal/services/mssql/mssql_database_resource.go:590]: I've confirmed in the SDK, but I'm really surprised this isn't a future - do we need to poll to wait for this to finish replicating at all?

**Review by tombuildsstuff — APPROVED** [review]
Left a few minor comments inline but this is otherwise G2G 👍

File: internal/services/mssql/mssql_database_resource.go (hunk 3 above)
Comment (koikonom) [reply internal/services/mssql/mssql_database_resource.go:590]: We'll now wait until the database has stopped scaling.

File: internal/services/mssql/mssql_database_resource.go [hunk 4]
```diff
@@ -558,6 +581,37 @@ func resourceMsSqlDatabaseCreateUpdate(d *pluginsdk.ResourceData, meta interface
...
+			TransparentDataEncryptionProperties: &sql.TransparentDataEncryptionProperties{
+				Status: statusProperty,
+			},
+		})
+		if err != nil {
+			return fmt.Errorf("while enabling Transparent Data Encryption for %q: %+v", id.String(), err)
+		}
+
+		if err = pluginsdk.Retry(d.Timeout(pluginsdk.TimeoutCreate), func() *pluginsdk.RetryError {
```
Comment (tombuildsstuff) [inline internal/services/mssql/mssql_database_resource.go:599]: fwiw you'll need to use the direct Plugin SDK import to remove this linter warning, we should probably file a bug/fix that

**Review by tombuildsstuff — APPROVED** [review]
one comment about the linting but otherwise 👍

**Pull request merged**